modules/currency-l1/src/main/scala/org/tessellation/currency/l1/http/DataApplicationRoutes.scala

```diff
@@ -47,64 +52,135 @@ final case class DataApplicationRoutes[F[_]: Async: Hasher: SecurityProvider: L1
-      req
-        .asR[Signed[DataUpdate]] {
-          _.toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _).flatMap {
-            case Left(_) => BadRequest(InvalidSignature.toApplicationError)
-            case Right(hashed) =>
-              validate(hashed.signed) {
-                dataUpdatesQueue.offer(hashed.signed) >> Ok(("hash" ->> hashed.hash.value) :: HNil)
+      def parseMultipleDataTransactions: F[DataTransactions] = {
+        implicit val dataTransactionRequest: EntityDecoder[F, DataTransactions] = dataApplication.requestDecoder
+        req.as[DataTransactions]
+      }
+
+      def parseSingleDataTransaction: F[DataTransactions] = {
+        implicit val decoder: EntityDecoder[F, Signed[DataUpdate]] = dataApplication.signedDataEntityDecoder
+        req
+          .as[Signed[DataUpdate]]
+          .map(NonEmptyList.of(_))
+      }
+
+      def handleDataUpdates(dataTransactions: DataTransactions, isSingleUpdate: Boolean): F[Response[F]] =
+        validateDataTransactions(dataTransactions) {
+          dataTransactions.traverse { signedTransaction =>
+            signedTransaction.value match {
+              case dataUpdate: DataUpdate =>
+                Signed(dataUpdate, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](dataApplication.serializeUpdate _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+
+              case feeTransaction: FeeTransaction =>
+                Signed(feeTransaction, signedTransaction.proofs)
+                  .toHashedWithSignatureCheck[F](serialize[F] _)
+                  .asInstanceOf[F[Either[InvalidSignatureForHash[DataTransaction], Hashed[DataTransaction]]]]
+            }
+          }.flatMap { results =>
+            val errors = results.collect { case Left(_) => InvalidSignature.toApplicationError }
+            val validTransactions = results.collect { case Right(hashed) => hashed }
+
+            if (errors.nonEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("Invalid signature in data transactions"))
+              )
+            } else if (validTransactions.isEmpty) {
+              BadRequest(
+                Json.obj("error" -> Json.fromString("No valid data transactions found"))
+              )
+            } else {
+              NonEmptyList.fromList(validTransactions.map(_.hash)) match {
+                case Some(hashes) =>
+                  dataTransactionsQueue
+                    .offer(NonEmptyList.fromListUnsafe(validTransactions.map(_.signed)))
+                    .flatMap { _ =>
+                      if (isSingleUpdate) {
+                        Ok(("hash" ->> hashes.head.value) :: HNil)
+                      } else {
+                        Ok("hashes" ->> hashes.map(_.value) :: HNil)
+                      }
```

Comment:
We shouldn't return an array of hashes when the hashes represent different things. There won't be anywhere in the system that we can look up a "DataTransaction" by hash, it will be either a DataUpdate or a FeeTransaction. If we return both of these in a single array then the client is forced to guess. 

I would suggest:
{
  "hash": "",
  "feeHash": ""
}

Reply:
Ok, I agree, as long as the input would be separated too. I thought we would send it merged. Otherwise, you're right, we should return data update hashes and fee txs hashes separately, but still as an array, not a combined string.

The response that you've just proposed does not scale if we allow to send multiple data updates / fee txs. IMO it should be an array of data hashes and fee hashes, not a single hash (single string) (because we allow to send a list of data updates, not a single data update).

Reply:
👍🏼